### ISSUE TITLE
'pip install triton' from pinned hash gives unreliable triton

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#100116]
Needs further investigation, but several users have reported that the following comand, which is supposedly fine to use, produces triton builds that sometimes segfault with pytorch+benchmarks.

`pip install "git+https://github.com/openai/triton@7d1a95b04654ff9c216afe08a454ad0822f05370#subdirectory=python"`
(replace 7d1a... with any valid pinned hash)

Instead, the `make triton` command which calls [install_triton_wheel.sh](https://github.com/pytorch/pytorch/blob/main/scripts/install_triton_wheel.sh) installs from our CI's build of triton, and this works reliably.

`pip install --extra-index-url https://download.pytorch.org/whl/nightly/ "pytorch-triton==$(cat .ci/docker/triton_version.txt)+$(head -c 10 .ci/docker/ci_commit_pins/triton.txt)"`

The difference is not the triton hash version as the same pinned hash was used with the working/broken install.

We should figure out if there is a missing or wrong config in the setup script for the regular pip build and make sure it works.

cc @malfet @seemethere @ezyang @soumith @msaroufim @ngimel @bdhirsh

**Comment by malfet** [comment]
Do you have any examples of segfaults? 

But I guess problem here, is a local system setup: `pip install --index-url https://download.pytorch.org/whl/nightly/ "pytorch-triton==$(cat .ci/docker/triton_version.txt)+$(head -c 10 .ci/docker/ci_commit_pins/triton.txt)"` installs binary that was build in the container that has compatible version of python, pybind, cmake, llvm, etc, while `pip install "git+https://github.com/openai/triton@7d1a95b04654ff9c216afe08a454ad0822f05370#subdirectory=python"` builds triton from source, using all the above dependencies from local users setup, which might result in incompatible dependencies, as many dev system have multiple version of Python, pybind, cmake installed on the system. Another difference, is that binary build links with llvm statically, which hides all the symbols, while local build would link with llvm dynamically, which can conflict with llvm PyTorch is built with or SciPy.

**Comment by ezyang** [comment]
At least in my experience, the problem was Triton needs to bootstrap against LLVM, which it has binaries for, but those binaries didn't work on the particular system I was compiling Triton on. The PyTorch binaries have all consistent compiler toolchain all the way through.

**Comment by bdhirsh** [comment]
I got a stack trace, in case it's useful (had to turn off inductor async compile). It's somewhere in `init_triton_ir()`:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c4d859 in __GI_abort () at abort.c:79
#2  0x00007ffff7cb826e in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7de2298 "%s\n") at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff7cc02fc in malloc_printerr (str=str@entry=0x7ffff7de05f5 "realloc(): invalid pointer") at malloc.c:5347
#4  0x00007ffff7cc60ba in __GI___libc_realloc (oldmem=0x7ffdcb70a5a8 <vtable for (anonymous namespace)::CustomOpAsmParser+16>, bytes=23923997776) at malloc.c:3175
#5  0x00007ffdc6f0495f in llvm::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#6  0x00007ffdc709143a in mlir::ParseResult llvm::function_ref<mlir::ParseResult ()>::callback_fn<mlir::AsmParser::parseTypeList(llvm::SmallVectorImpl<mlir::Type>&)::{lambda()#1}
>(long) () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#7  0x00007ffdc931455a in mlir::detail::Parser::parseCommaSeparatedList(mlir::AsmParser::Delimiter, llvm::function_ref<mlir::ParseResult ()>, llvm::StringRef) [clone .localalias]
 () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#8  0x00007ffdc743393f in mlir::AsmParser::parseCommaSeparatedList(llvm::function_ref<mlir::ParseResult ()>) (this=<optimized out>, parseElementFn=...)
    at /data/home/hirsheybar/.triton/llvm/llvm+mlir-17.0.0-x86_64-linux-gnu-ubuntu-18.04-release/include/mlir/IR/OpImplementation.h:701
#9  mlir::AsmParser::parseTypeList (this=<optimized out>, result=...)
    at /data/home/hirsheybar/.triton/llvm/llvm+mlir-17.0.0-x86_64-linux-gnu-ubuntu-18.04-release/include/mlir/IR/OpImplementation.h:1167
#10 mlir::triton::ExtElemwiseOp::parse (parser=..., result=...)
#11 0x00007ffdc932228a in (anonymous namespace)::OperationParser::parseOperation() () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton[42/1481]
#12 0x00007ffdc9325fbd in (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#13 0x00007ffdc93266a1 in (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
--Type <RET> for more, q to quit, c to continue without paging--
  py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#14 0x00007ffdc9326855 in (anonymous namespace)::CustomOpAsmParser::parseOptionalRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#15 0x00007ffdc76e1712 in mlir::function_interface_impl::parseFunctionOp(mlir::OpAsmParser&, mlir::OperationState&, bool, mlir::StringAttr, llvm::function_ref<mlir::Type (mlir::B
uilder&, llvm::ArrayRef<mlir::Type>, llvm::ArrayRef<mlir::Type>, mlir::function_interface_impl::VariadicFlag, std::__cxx11::basic_string<char, std::char_traits<char>, std::alloca
tor<char> >&)>, mlir::StringAttr, mlir::StringAttr) () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#16 0x00007ffdc70a76ea in mlir::func::FuncOp::parse(mlir::OpAsmParser&, mlir::OperationState&) ()
    at /data/home/hirsheybar/.triton/llvm/llvm+mlir-17.0.0-x86_64-linux-gnu-ubuntu-18.04-release/include/mlir/IR/Operation.h:829
#17 0x00007ffdc932228a in (anonymous namespace)::OperationParser::parseOperation() () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#18 0x00007ffdc9325fbd in (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#19 0x00007ffdc93266a1 in (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#20 0x00007ffdc9326731 in (anonymous namespace)::CustomOpAsmParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::Argument>, bool) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#21 0x00007ffdc76885b7 in mlir::ModuleOp::parse(mlir::OpAsmParser&, mlir::OperationState&) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#22 0x00007ffdc932228a in (anonymous namespace)::OperationParser::parseOperation() () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#23 0x00007ffdc9323d32 in mlir::parseAsmSourceFile(llvm::SourceMgr const&, mlir::Block*, mlir::ParserConfig const&, mlir::AsmParserState*, mlir::AsmParserCodeCompleteContext*)
    () from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#24 0x00007ffdc92f8154 in mlir::parseSourceFile(std::shared_ptr<llvm::SourceMgr> const&, mlir::Block*, mlir::ParserConfig const&, mlir::LocationAttr*) [clone .localalias] ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#25 0x00007ffdc92f8793 in mlir::parseSourceFile(llvm::StringRef, mlir::Block*, mlir::ParserConfig const&, mlir::LocationAttr*) ()
   from /scratch/hirsheybar1/work/py38/lib/python3.8/site-packages/triton/_C/libtriton.so
#26 0x00007ffdc6b5a841 in mlir::detail::parseSourceFile<mlir::ModuleOp, llvm::StringRef&> (config=..., args=...)
    at /data/home/hirsheybar/.triton/llvm/llvm+mlir-17.0.0-x86_64-linux-gnu-ubuntu-18.04-release/include/mlir/Parser/Parser.h:159
#27 mlir::parseSourceFile<mlir::ModuleOp> (filename=..., config=...)
    at /data/home/hirsheybar/.triton/llvm/llvm+mlir-17.0.0-x86_64-linux-gnu-ubuntu-18.04-release/include/mlir/Parser/Parser.h:203
#28 init_triton_ir(pybind11::module_&&)::$_28::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mlir::MLIRContext&) const (
    this=<optimized out>, inputFilename=..., context=...) at /tmp/pip-req-build-_58m9wgw/python/src/triton.cc:408
```

**Comment by hiroshinoji** [comment]
I encountered a similar issue when using a triton compiled locally, and for me, it was fixed by changing g++ version to 9 when compiling triton. The default GCC version was 8 in my environment (ubuntu 18.0.4). For me, the cause looks to be in linking to std::filesystem, specifically, this line looks to be failed.
 https://github.com/openai/triton/blob/2689f4a3b01a60a4fb6ec92d94f46e5948682807/lib/Target/LLVMIR/LLVMIRTranslation.cpp#L192

Another way might be adding `-lstdc++fs` explicitly, though that didn't help in my case.
